### PR TITLE
Improve Paper UI, allow an to entity configure how it's UI looks

### PR DESCRIPTION
**Pull request by eoineoineoin**
## About the PR

Improves the paper UI - crew are no longer writing on slate tiles, and stamps are visible on the UI.

Adds a component which can be added to an entity which controls how the paper is rendered on the client, including several samples (in screenshots below) - I'm not particularly attached to any of these, if people think this is too much (or can do better art) - and update some machines to produce department-thematic paper. Books get a nice UI as a bonus, too.

When sending paper with a fancy UI, fax machines will preserve the style. (I guess you can fax whole books, actually?)

I think this can still be taken a bit further, but in the interests of keeping this PR shorter, will leave them for later:
* There's loads of opportunity for new paper types (empty chocolate wrappers? whiteboards? ticket/queueing machine? character sheets, you name it)
* Stamps could have different styles & ink color (would be great to get more ability to composite UIs)
* Different pens could write in different fonts

I also think the vending machine could use a similar makeover, so may want to pull out some of these into a more generic component. 

Couple of notes:

I'm wasn't sure where the best place to put the images was - I settled on Resources/Textures, since that was where the UI textures seemed to go and using an RSI wasn't really suitable, as the textures used as building blocks for these effects tend to be different sizes.

I'm not super happy about the change to cargo system, but I see #13444, so I will probably tackle that soon, if nobody else is doing it?

I'm seeing a unit test failure due to a missing file Textures/Interface/Alerts/human_health.rsi, - not something I changed (intentionally) so may need to dig into that.

I see #13449 is in progress - might be neat to also spruce up the kinds of paper that those bins offer.

**Media**

First, the paper UI as it is on the main branch:

original

"Plain" paper - the UI you get if you have a PaperComponent, but have not added PaperVisualsComponent (also, the default; this can be controlled by the style, too)

plain_paper

For serious office work, you need lined paper:

lined_plain

The rendering takes into account the user's/style's font and UI scale preferences to ensure that crew members always write between the lines, so there's no point reading between them (true for all the UIs, not just this one):

lined_scaled_2_2023-01-15_19-08
lined_scaled_2023-01-15_19-04

Cargo order manifests are now produced on yellow carbon-copy and now clearly state the originating department:

cargo_invoice

The disease diagnoser prints out little receipts instead of A4 sheets, helping to reduce waste:

diagnosis_machine

The artifact analyzer's printer is out of warranty:

artifact_analyzer

And the default book looks like you casually opened it on the right page (the only page):

book

Captain has a dedicated pad for writing important notes (is this too silly?)

captain_2

Not included in this PR, but some tests I made to see if there's room for adding flavour, and shows some extra capabilities of the settings:

fun_businesscard
fun_napkin

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Cripes
- add: Nanotrasen has updated the available types of paper in the stationary department
- add: Disease diagnoser, artifact analyzer and cargo shuttle produce themed paper

### CODE REVIEW

File: Content.Client/Paper/UI/PaperBoundUserInterface.cs [hunk 1]
```diff
@@ -19,14 +19,21 @@ public PaperBoundUserInterface(ClientUserInterfaceComponent owner, Enum uiKey) :
...
             _window = new PaperWindow
             {
-                Title = IoCManager.Resolve<IEntityManager>().GetComponent<MetaDataComponent>(Owner.Owner).EntityName,
+                Title = entityMgr.GetComponent<MetaDataComponent>(Owner.Owner).EntityName,
             };
             _window.OnClose += Close;
             _window.Input.OnTextEntered += Input_OnTextEntered;
-            _window.OpenCentered();
 
+            if (entityMgr.TryGetComponent<PaperVisualsComponent>(Owner.Owner, out var visuals))
```
Comment (ZoldorfTheWizard) [inline Content.Client/Paper/UI/PaperBoundUserInterface.cs:31]: *.Owner is deprecated if i recall correctly from mr sloth, pass the owner in via method arguments

**Review by ZoldorfTheWizard — CHANGES_REQUESTED** [review]

**Comment by eoineoineoin** [comment]
> That's really neat! Can the paper expand down if the text is longer than the default size?

It does, if the paper type permits it. Say, the diagnosis receipt is printed on a roll so can expand vertically, but not horizontally (so this is almost physically realistic, I suppose):

https://user-images.githubusercontent.com/1676295/212571706-734426d3-60d5-4c90-b237-003ac6a8e563.mp4

and the business card sample in the initial PR just has a _teeny_ max area and can't be resized (it's the bit in the bottom right that says "Call me - Dirk") -- however, normal non-gimmick paper doesn't have any restrictions on being resized. This can still result in some weird situations if the user makes the window too small and I had to put in a default MaxWidth (which can still be overridden in the PaperVisualsComponent) as some pre-written books had long lines and I couldn't see a way to keep a reasonable size I was happy with:



A limit the size of all paper, so that it always has a max area? Maybe, but probably not fun.

Using a ScrollBoxContainer when the content region is too small? Seemed a little weird to add a scrollbox to paper (though could probably handwave some space paper magic around here)

File: Content.Client/Paper/UI/PaperBoundUserInterface.cs (hunk 1 above)
Comment (EmoGarbage404) [reply Content.Client/Paper/UI/PaperBoundUserInterface.cs:31]: there isn't a replacement for this yet AFAIK.

**Comment by AJCM-git** [comment]
> A limit the size of all paper, so that it always has a max area? Maybe, but probably not fun.
> Using a ScrollBoxContainer when the content region is too small? Seemed a little weird to add a scrollbox to paper (though could probably handwave some space paper magic around here)

Supporting a max area and scrolling is kinda important since i can see some paper designs wanting to remain thin and going out of the screen with a long text if there is no max area

**Comment by arealbeba** [comment]
Maybe you can make stamp colors match on the paper? Like captain's stamp is going to be shown blue on the paper.

**Comment by eoineoineoin** [comment]
> > A limit the size of all paper, so that it always has a max area? Maybe, but probably not fun.
> > Using a ScrollBoxContainer when the content region is too small? Seemed a little weird to add a scrollbox to paper (though could probably handwave some space paper magic around here)
> 
> Supporting a max area and scrolling is kinda important since i can see some paper designs wanting to remain thin and going out of the screen with a long text if there is no max area (and if there is a max area, a scrollbar is kind of required)

OK, I've done a little bit of experimenting with the scroll container and was able to get something that behaves well, especially when there is a lot of content, or the window has been resized to something small.

There was also a suggestion on Discord to remove the window background, to get the full effect of the paper: 


_Immersive_.

This also gets the nice behavior that I can allow/prohibit resizing on X/Y individually.
I tried putting the close button inside the paper, didn't feel as nice, and sometimes could overlap written content.

**Comment by eoineoineoin** [comment]
> Maybe you can make stamp colors match on the paper? Like captain's stamp is going to be shown blue on the paper.

I would actually go the other way, here. The stamp sprites are colour-coded, so it would be neat if each stamp passed along it's colour.

File: Content.Client/Paper/UI/PaperBoundUserInterface.cs (hunk 1 above)
Comment (Jezithyr) [reply Content.Client/Paper/UI/PaperBoundUserInterface.cs:28]: BUI's need a refactor sometime >.>

File: Content.Client/Paper/UI/PaperVisualsComponent.cs [hunk 2]
```diff
@@ -1,6 +1,71 @@
...
 public sealed class PaperVisualsComponent : Component
 {
+    /// The path to the image which will be used as a background for the paper itself
+    [DataField("backgroundImagePath")]
+    public string? BackgroundImagePath;
+
+    /// An optional patch to configure tiling stretching of the background. Used to set
+    /// the PatchMargin in a `StyleBoxTexture`
+    [DataField("backgroundPatchMargin")]
+    public Box2 BackgroundPatchMargin = default;
```
Comment (metalgearsloth) [inline Content.Client/Paper/UI/PaperVisualsComponent.cs:13]: These should be xml docs.

File: Content.Server/Fax/FaxSystem.cs [hunk 3]
```diff
@@ -397,12 +398,17 @@ public void Send(EntityUid uid, FaxMachineComponent? component = null, EntityUid
             { FaxConstants.FaxPaperContentData, paper.Content },
         };
 
+        if (metadata.EntityPrototype != null)
+        {
+            payload[FaxConstants.FaxPaperPrototypeData] = metadata.EntityPrototype.ID;
+        }
+
```
Comment (metalgearsloth) [inline Content.Server/Fax/FaxSystem.cs:405]: This should just copy visuals instead; if an entity is constructed entirely ingame / in code vs prototype it should function the same and not rely on hardcoding prototype ID copies.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Fax/FaxSystem.cs (hunk 3 above)
Comment (metalgearsloth) [reply Content.Server/Fax/FaxSystem.cs:405]: Alright having talked to Jez archetypes has entity copies which is what it should do instead so this just needs a TODO to not do this ID check anymore.

**Review by metalgearsloth — APPROVED** [review]

**Comment by eoineoineoin** [comment]
Wow, that was a low time between pushing and merging! Thanks for reviewing and providing those tips.

**Pull request merged**